[ARM] Emit an error when the hard-float ABI is enabled but can't be used

chrisnc opened the pull request:
Prior to this, compiling for an eabihf target with a CPU lacking
floating-point registers would silently use the soft-float ABI instead,
even though the Arm attributes section would still have
"Tag_ABI_VFP_args: VFP registers", which leads to silent ABI mismatches
at link time.

Update various tests that were using inconsistent ABI/PCS and features.

Change ARMTargetLowering::getEffectiveCallingConv from private to public
and modify it to pass through unrecognized calling conventions. Now that
ARMBaseTargetMachine::createMachineFunctionInfo calls it, leaving a
fatal error would change the behavior of IR passes that do not need to
lower calling conventions. Unrecognized calling conventions are still
treated as errors in lowering.

Fixes #110383.

chrisnc commented:
clang tests are passing now, but some llvm tests are not, e.g., invocations of llc in `CodeGen/ARM/arm-eabi.ll` that specify e.g., `--target=arm-none-eabihf` don't include enough features to actually use eabihf. Any suggestions on the right place to address this would be appreciated.

davemgreen commented:
It looks like there is already a warning for this in clang, it only triggers from -mfloat-abi=hard though: https://godbolt.org/z/dcaz8had4. Could it be made to work with any hard-float env? And maybe be made an error down-gradable to a warning?

Generally clang-level warnings/errors are more user friendly then the llvm-level errors (but both may be useful for other frontends).

chrisnc commented:
> It looks like there is already a warning for this in clang, it only triggers from -mfloat-abi=hard though: https://godbolt.org/z/dcaz8had4. Could it be made to work with any hard-float env? And maybe be made an error down-gradable to a warning?
> 
> Generally clang-level warnings/errors are more user friendly then the llvm-level errors (but both may be useful for other frontends).

That seems like a decent improvement to the clang warning, but the backend behavior of still emitting the VFP ABI tag in this case seems like a bug, regardless of the frontend, so IMO codegen should be aborted in this situation.

smithp35 commented:
Whatever we choose to do I think clang and llvm should be consistent. There's no point in clang giving a warning if llvm is going to give an error.

I can see two positions most clearly:
* clang gives an error message for an eabihf target (or -mfloat-abi=hard). This is what GCC does today. While we could also put an error in LLVM I'm not sure it is necessary.
* clang gives a warning message for both eabihf target (of -mfloat-abi=hard) and uses -mfloat-abi=soft, the warning should ideally say that is changing ABI rather than just unsupported. We'd need to change LLVM to get the build attributes right.

Personally I don't have a problem with clang giving an error in this case as that is what GCC does.

Appendix, some empirical data:
GCC 13
```
arm-none-eabi-gcc -c -mcpu=cortex-m3 -mfloat-abi=hard float.c
cc1: error: '-mfloat-abi=hard': selected architecture lacks an FPU
```
Clang -mfloat-abi
```
clang --target=arm-none-eabi -mcpu=cortex-m3 float.c -c -mfloat-abi=hard
clang: warning: '-mfloat-abi=hard': selected processor lacks floating point registers [-Wunsupported-abi]
```
Actual behaviour is soft-float but with incorrect build attributes.

Clang --target=arm-none-eabihf
```
clang --target=arm-none-eabi -mcpu=cortex-m3 float.c -c -mfloat-abi=hard
```
Actual behaviour is soft-float but with incorrect build attributes.

chrisnc commented:
Among those options, my preference would be to remove the clang warning and add the LLVM error so that other frontends can benefit from this checking, as well as the LLVM tools themselves. The LLVM test failures reveal a few places where this issue seems to be ignored.

RalfJung commented:
> Generally clang-level warnings/errors are more user friendly then the llvm-level errors (but both may be useful for other frontends).

What would be really nice is if the LLVM backend could tell the frontend about the error rather than reporting it itself, that way frontends can use their nice error reporting code but don't have to re-implement the error detection logic. :)

Anyway, that's a different issue...

chrisnc commented:
I've updated this PR to remove the aforementioned `clang` warning for this case, and I've updated all `clang` and `llvm` tests that were implicitly using the soft-float ABI despite requesting hard-float. In some cases, explicitly testing hard-float seemed to be intended because there was a corresponding line for the soft-float version, so I added a feature or CPU selection that accomplishes this. In other cases, the hard-float-ness of the test did not seem relevant, so I just switched those to soft-float. I attempted to write a test case for the new LLVM error, but I'm struggling with having a `lit` test that tolerates `llc` exiting with a nonzero exit code, even if I can get the `CHECK` directives to match. Any pointers on this would be appreciated.

chrisnc commented:
All unit tests are now updated and passing and this change is ready for review. Thanks!

chrisnc commented:
Friendly bump :) @davemgreen @smithp35

smithp35 commented on llvm/lib/Target/ARM/ARMTargetMachine.cpp:
```diff
@@ -311,11 +311,15 @@ ARMBaseTargetMachine::getSubtargetImpl(const Function &F) const {
     if (!I->isThumb() && !I->hasARMOps())
       F.getContext().emitError("Function '" + F.getName() + "' uses ARM "
           "instructions, but the target does not support ARM mode execution.");
+
```
IIUC [*] TargetMachine::getSubtargetImpl is the "global" subtarget. It has been a long time since I've looked at the area so I could be wrong.

Putting the warning here makes me nervous as individual functions have their own subtargets. It is common for LTO to rely entirely on the per function subtarget. I can remember an Android problem that was caused by using the global subtarget for a per-function property.

chrisnc replied:
`getSubtargetImpl` takes a `const Function &` as a parameter, and looks at that function's attributes, and can give per-function diagnostics for the thumb check, so I believe it is not the "global" subtarget.

smithp35 commented on clang/include/clang/Basic/DiagnosticDriverKinds.td:
```diff
@@ -459,9 +459,6 @@ def warn_drv_assuming_mfloat_abi_is : Warning<
 def warn_drv_unsupported_float_abi_by_lib : Warning<
   "float ABI '%0' is not supported by current library">,
   InGroup<UnsupportedABI>;
-def warn_drv_no_floating_point_registers: Warning<
```
This is a build break for anyone that is explicitly referring to this warning.

I'm not sure what the protocol for removing warnings is in clang. For example are they release noted/deprecated or just left in and ignored?

chrisnc replied:
`-Wunsupported-abi` still exists (it is a group containing several warnings). The only thing that is removed here is the particular manifestation of this warning for the ARM target. It should only be a build breakage if `-Wunsupported-abi` were removed completely.

smithp35 reviewed (commented):
I'm not yet convinced that llvm is the right place for this error message. There's quite a lot of test changes that I presume needed to make and at least the LTO use case looks like we don't want to require extra information. I'd also prefer this to be a front-end error/warning as this is a front-end/source-level mistake, with better diagnostic control options available.

I think if it is implemented in llvm then it will need to use the per-function subtarget to not risk fallout with LTO builds.

Apologies I'm probably not particularly enthusiastic about the approach. I've not worked in this area for some time so I'm happy for other reviewer's opinions to take precedence.

RalfJung commented:
> I'm not yet convinced that llvm is the right place for this error message. There's quite a lot of test changes that I presume needed to make and at least the LTO use case looks like we don't want to require extra information. I'd also prefer this to be a front-end error/warning as this is a front-end/source-level mistake, with better diagnostic control options available.

There's a *lot* of complicated logic required to determine, for each architecture, which target features affect the ABI in which ways. Are you suggesting that logic should be duplicated across all frontends? That seems like a huge waste of effort to me. Most frontends are inevitably going to get it subtly wrong, so in the end we have a ton more bugs than we did if there was a central location where such know-how could be properly encoded, and where there are enough people that can confidently answer ABI questions like this.

smithp35 commented:
> > front-end error/warning as this is a front-end/source-level mistake, with better diagnostic control options available.
> 
> There's a _lot_ of complicated logic required to determine, for each architecture, which target features affect the ABI in which ways. Are you suggesting that logic should be duplicated across all frontends? That seems like a huge waste of effort to me. Most frontends are inevitably going to get it subtly wrong, so in the end we have a ton more bugs than we did if there was a central location where such know-how could be properly encoded, and where there are enough people that can confidently answer ABI questions like this.

It is a tradeoff. Yes it means redoing effort in front-ends, but it can mean a better experience for the end-user, and to me that is often more important than the work that we do as toolchain developers. I'm not saying I've definitely got the balance right in this case which is why I invite more opinions :-)

An error in the backend does not have a lot of context. In some cases it may end up triggering on a large LTO build without any source file information given in the error message.

RalfJung commented:
Yeah ideally LLVM would provide APIs where frontends can explicitly query that kind of information, and then provide nice errors with context. What this PR does is only the second-best think, in my eyes.

But very few frontends can afford having an expert for each and every architecture LLVM supports, so "do it in the frontend" is often simply not an option.

chrisnc commented:
> There's quite a lot of test changes that I presume needed to make

FWIW, virtually all of the test changes were due to cases where LLVM would silently use soft-float despite the hard-float ABI being explicitly requested, which is exactly the incorrect behavior that this PR is meant to fix. The total number of affected test cases is fairly small.

> and at least the LTO use case looks like we don't want to require extra information.

The direct invocation of llvm-lto case without target information is one I will look into, however, this change was made again because llvm was generating code in a manner where the hard-float ABI was requested but couldn't be fulfilled. Anywhere that might be happening can lead to unsound ABI-mixing. To me the question is whether the equivalent of that command is invoked internally without the correct target information in typical usage of LTO driven by clang/lld. I'm inclined to say no, based on there being zero other test failures related to this change, but I will do some digging to understand the situation.

chrisnc commented:
@ostannard or @davemgreen any other feedback on this change?

ostannard commented:
I agree with the previous comments from @davemgreen and @smithp35, user-facing diagnostics should always be emitted by clang, so that cases like LTO work correctly. If you want to reduce the duplication of the code which works out what ABIs are compatible (which I think would be a good idea), you could probably put that code in `TargetParser`, and use it from clang to decide what error/warning to emit.

davemgreen commented:
I agree but was having trouble putting it into words. I don't have a reference I can put my hands on, but we have generally considered the llvm error messages to be a poor substitute to those produced by clang and have often gone the other way, converting llvm errors to clang errors.  I would think the best interface here is a error/warning that is downgradable to a warning/silent and that is difficult to do in llvm at the moment with the way the errors are setup. They are mostly reserved for things that have gone very wrong or cannot be supported.

It would be good to share this kind of logic where we can though. Either by building a proper way for the backends to report errors to the frontends, or for the logic to be shared in TargetParser.

RalfJung commented:
As a Rust compiler dev, this part of LLVM feels extremely fragile. If we do anything wrong, LLVM will just use *some* ABI, and things will keep working for some situations, but really we have subtle ABI bugs that will eventually bite our users. The rules we have to follow are not documented anywhere (to my knowledge), and require target-specific expertise -- even if we get the x86 part right, we'll need a PowerPC expert to also get soft-float handling right for that target.

It's not a pleasant experience at all. Having LLVM be more robust, and throw errors rather than being like "yeah let's just continue, I'm sure it's going to be fine", would go a long way towards making LLVM easier to use and more reliable as a backend. It is concerning if attempts at improving the robustness of LLVM in this space are being blocked.

smithp35 reviewed (commented):
Apologies for the delay in responding.

I don't have much to add over davemgreen and ostannard.

Leaving aside whether this should be an error in the front or back-end. As mentioned before I haven't worked in this area for a long time, and there are enough test changes here that make me nervous enough that this is going to break some existing use cases. Some do look like sloppy triple selection, although a lot of the tests don't use any floating point so it could be seen as harsh to fault them. If I understood the area better that wouldn't be a problem, but as I'm not enthusiastic about moving a diagnostic to the backend, this isn't a change that I am best placed to approve. I won't object if there's a consensus in the other direction.

One use case I'd want to keep working is where we have two functions in the same object that are hardfp with hardware registers, and soft (with or without hardware registers). Something like:
```
__attribute__((target("arch=cortex-m33"))) __attribute__((pcs("aapcs-vfp"))) float f1(float a, float b) {
  return a + b;
}

__attribute__((target("arch=cortex-m0"))) __attribute__((pcs("aapcs"))) float f2(float a, float b) {
  return a + b;
}
```
There are some source files that do run-time selection based on hardware detection.

chrisnc commented on clang/test/Driver/arm-float-abi-lto.c:
```diff
@@ -16,7 +16,7 @@
 // RUN: %clang --target=arm-none-eabi -mcpu=cortex-m33 -mfloat-abi=hard -O1 %s -flto=thin -c -o %t.call_thin.bc -DCALL_LIB
 // RUN: %clang --target=arm-none-eabi -mcpu=cortex-m33 -mfloat-abi=hard -O1 %s -flto=thin -c -o %t.define_thin.bc -DDEFINE_LIB
-// RUN: llvm-lto2 run -o %t.lto_thin -save-temps %t.call_thin.bc %t.define_thin.bc \
+// RUN: llvm-lto2 run --mcpu=cortex-m33 --float-abi=hard -o %t.lto_thin -save-temps %t.call_thin.bc %t.define_thin.bc \
```
I dug into this further as promised, and the error goes away for this test if you just remove the call to `printf` from the test, which the test expects to replace with a call to putchar. This means that the `llvm-lto2` invocation is eventually creating a new subtarget in which the target features and ABI are not inherited from the settings of either input file, and because the ones it lands on are incompatible, an error is raised. Pre-existing functions that were generated with a specific ABI and features are preserved as you would expect.

chrisnc replied:
To test this theory, I had the error output the function name just like the nearby check for thumb does, and this is what I get when running this test unmodified:

`The hard-float ABI is enabled for function 'llvm.lifetime.start.p0', but the target lacks floating-point registers.`

This means that this function was being generated during LTO itself with an invalid target+ABI combination. Depending on what this functions parameters are, that could lead to incorrect behavior, so in general, it may not be safe to LTO without enough features enabled to support the ABI that the compiled inputs use, and it is good for LLVM to complain about this rather than potentially emit erroneous code.

smithp35 replied:
Thanks for the investigation. I think that this use case of LTO needs to work without an error as while we can provide the target for this specific test case, it may cause other uses of LTO without a target on the linker driver to fail. This may mean that we have to fix this in LTO.

From git annotate the test case was added in https://reviews.llvm.org/D89573 ostensibly to make sure that if LTO calls a C function it uses the correct ABI.

Looking at llvm.lifetime.start this is a lifetime marker intrinsic https://llvm.org/docs/LangRef.html#llvm-lifetime-start-intrinsic . IIUC lifetime markers are for communicating lifetimes of objects to optimisers, so while they influence code-generation they don't get code generated for them.

It may be an area where the subtarget for the function isn't being passed through correctly, although in the case of lifetime markers it is benign. This is an area that needs a lot more in depth expertise of LLVM and LTO than I posess. I can ask around internally to see if anyone is willing to take a look.

chrisnc replied:
What do you think about making the error only fire if any of the function's argument types or return types are floating-point? That would eliminate the need for this test change and probably several others that weren't actually using floating-point, and be more defensible in that if you encounter such a function with the ABI/feature combination being checked for, it _will_ generate incorrect code for it if not for this error.

smithp35 replied:
I think that would be a good way forward. That would help reduce the chances of this breaking some existing projects.

That would then leave us with the debate about front-end and back-end error messages. I've been thinking about how to get past the differences of opinion here. My thoughts so far:
* Make the clang warning an error (to match GCC) and make sure it triggers for the target and the -mfloat-abi=hard case. Then clang users will see a front-end error and shouldn't see the backend error unless they've messed about behind the compilers back with target attributes. Other front-ends can choose their own policy.
* Make the back-end error opt-in, with clang opting out and retaining the existing clang warning. That retains the existing behaviour for clang. I personally would prefer clang/gcc to behave the same way and give an error message.

While I think that either of those would take care of user facing concerns, there's a possible objection of code complexity, but I'd have to leave that in the hands of the code-owners.

chrisnc replied:
I attempted this approach, but there's an issue due to how the `getSubtargetImpl` function works. It essentially maintains a map of subtargets it's seen before, keyed by a string of cpu + feature list, and only does the error checking when it sees a new combination, otherwise returns the subtarget that's already been created for the given cpu + feature list key. What this means is that if a non-fp-using function is encountered first in a translation unit, then the check passes because its use of inconsistent ABI+feature is harmless, but subsequent functions that should be flagged will be ignored because the map already has a subtarget for them.

Any suggestions on how to deal with this would be appreciated, or if there is a maintainer for this specific subsystem who could provide guidance here that would also be helpful.

RalfJung commented on llvm/lib/Target/ARM/ARMTargetMachine.cpp:
```diff
@@ -311,11 +311,15 @@ ARMBaseTargetMachine::getSubtargetImpl(const Function &F) const {
       F.getContext().emitError("Function '" + F.getName() + "' uses ARM "
           "instructions, but the target does not support ARM mode execution.");
+
+    if (I->isTargetHardFloat() && !I->hasFPRegs())
+      F.getContext().emitError("The hard-float ABI is enabled, but the target "
+                               "lacks floating-point registers.");
```
This seems to check the target triple and do automatic hard float detection. Shouldn't this instead check Options.FloatABIType?

chrisnc replied:
I think you are right. I initially avoided that because one of the values is `Default` which is meant to defer to the triple, but in the constructor of `ARMBaseTargetMachine` it sets the internal `Options.FloatABIType` to one or the other based on the passed in option and the triple, so it should be correct to just use that. I was testing with things like `--target=arm-none-eabi -mcpu=cortex-m0 -mfloat-abi=hard` and seeing the error trigger, so something else in the frontend may be generating a different triple that makes this "work".

chrisnc replied:
Ah, I remember now where I got stuck on this before. I need to obtain this property on the `ARMSubtarget` rather than the `ARMBaseTargetMachine`, but neither of them expose `Options` publicly, and `ARMSubtarget` just invokes `isTargetHardFloat` on its `TM` member, without accounting for its own `Options`, even though they are copied from the `ARMBaseTargetMachine` when the subtarget is created.

RalfJung replied:
Yeah `ARMSubtarget.isTargetHardFloat` seems like a footgun, I hope the rest of the backend properly has `Options.FloatABIType` override the target default.

RalfJung commented:
RISC-V has similar checks here:

https://github.com/llvm/llvm-project/blob/ed572f2003275da8e06a634b4d6658b7921e8334/llvm/lib/Target/RISCV/RISCVISelLowering.cpp#L88-L100

So maybe the ARM checks could be added in a similar place? RISC-V handles ABI variants in a very clean way, and it does seem to work in practice too, so it'd be a good model for other architectures to follow.

chrisnc commented:
> RISC-V has similar checks here:
> 
> https://github.com/llvm/llvm-project/blob/ed572f2003275da8e06a634b4d6658b7921e8334/llvm/lib/Target/RISCV/RISCVISelLowering.cpp#L88-L100
> 
> So maybe the ARM checks could be added in a similar place? RISC-V handles ABI variants in a very clean way, and it does seem to work in practice too, so it'd be a good model for other architectures to follow.

I could try this approach here and see how it goes.

RalfJung commented:
@chrisnc what are the next steps here?

Maybe one solution, for maximum compatibility, would be to not emit any warnings when `Options.FloatABIType` is left at `FloatABI::Default`, but if it is explicitly set to `FloatABI::Hard` and the required target features are missing, that should warn. (I assume `FloatABI::Soft` requires no target features.) `FloatABI::Hard` also seems to be incompatible with the `soft-float` target feature so that should probably also be a warning.

chrisnc commented:
I've been fairly busy over the past few months so haven't had time to pursue this further. One of the issues is that as far as I can tell, there doesn't seem to be a mechanism to issue a warning from within LLVM codegen, only to error. The behavior you are describing is already what `clang` does, but there isn't a way to just move that into LLVM so it can be easily used. There's also the matter of whether to treat an explicit choice of target that has eabihf as the ABI component in the same way. For this to be generally useful to Rust I think you have to treat the target choice as opting in to that ABI, but that is not the current behavior of the `clang` warning, and there are several test cases invoking the LLVM tools directly that expect the behavior of just not using the same ABI that the target says it should...

My next step was going to experiment with the approach the RISC-V backend uses and see how it goes, especially to evaluate the blast radius in the tests and compare.

RalfJung commented:
> there doesn't seem to be a mechanism to issue a warning from within LLVM codegen, only to error. 

RISC-V does `errs() << "Hard-float 'f' ABI can't be used for a target that "  ...` and that acts as a warning. It just prints a message to stderr. It's not pretty but it's better than generating code that has the wrong ABI and spending an hour debugging that...

Rust will always set `Options.FloatABIType` to a non-`Default` value for ARM-32.

chrisnc commented:
> RISC-V does `errs() << "Hard-float 'f' ABI can't be used for a target that " ...` and that acts as a warning. It just prints a message to stderr. It's not pretty but it's better than generating code that has the wrong ABI and spending an hour debugging that...

Oh, good to know... I specifically went to check where that interface was defined to see if there was an equivalent `warn()` IO stream or something before writing that comment, but hadn't triggered those checks myself before.

Even if Rust always sets a non-default ABI, it's not clear which use cases of clang and the llvm tools result in the same, but that will be clearer once I see the test runs.

I'll try and pick this up again this weekend.

chrisnc commented:
I've adapted the approach used in the RISC-V backend to this PR and carried over the new test I added before (updated the warning message and changed the expected exit code of the erroneous `llc` invocation to success instead of failure). From local testing, no other tests needed to be changed, but will check CI to confirm.

Would appreciate another review.

I've opted to leave the `clang` warning for this intact, as it deals with more narrow situations and may be more friendly, though there are cases where both will occur.

RalfJung commented:
That seems to implement what was sketched above, but I am not familiar with this code so I can't say if there's any possible problems this could cause.

chrisnc commented:
CI is passing with no changes to existing tests. @davemgreen @ostannard @smithp35 please let me know if any further changes are needed. Having this backstop will give the users of other tools a chance at avoiding these situations.

smithp35 commented on llvm/lib/Target/ARM/ARMISelLowering.cpp:
```diff
@@ -791,7 +791,9 @@ ARMTargetLowering::ARMTargetLowering(const TargetMachine &TM,
     if (!Subtarget->hasFP64())
       setAllExpand(MVT::f64);
-  }
+  } else if (TM.Options.FloatABIType == FloatABI::Hard)
```
When I run this on a test program I get a couple of warnings
> bin/clang --target=arm-none-eabihf -march=armv7-r nofp.c -c -mfloat-abi=hard -ffunction-sections
clang: warning: '-mfloat-abi=hard': selected processor lacks floating point registers [-Wunsupported-abi]
The hard-float ABI can't be used for a target that doesn't support floating-point (ignoring float-abi)
The hard-float ABI can't be used for a target that doesn't support floating-point (ignoring float-abi)

I think this is because this constructor is run once per ARMSubtarget. Tracing this back to getSubtargetImpl where ARMSubtargets are created https://github.com/llvm/llvm-project/blob/main/llvm/lib/Target/ARM/ARMTargetMachine.cpp#L280

It looks like that might be a better place to put this diagnostic. We may also want to see if some additional information like the function name can be used, possibly something from the Subtarget if the output is human readable.

ostannard replied:
That does look like a better place for this, and already has one check like this, emitted through the `LLVMContext` so that the compiler can exit with a non-zero status code, or report the error through whatever other means it normally used. I think an error is better than a warning for this, because it's not possible to generate the requested code correctly.

chrisnc replied:
The first version of the PR did put the check in that spot, but it leads to several required test changes because many tests invoke llc or lto with hardfp and no fp support, but then don't actually use fp, whereas the current patch does not result in those tests failing. Moving the check there means that those tests will need to be changed to invoke the LLVM tools with correct options. What is the reviewers' preference? @ostannard @smithp35

ostannard replied:
I think it would be best to update those tests, probably to use the the soft-float ABI if they don't actually use floating-point.

chrisnc replied:
The specific case of concern was this one, which is still ultimately due to not requesting the correct attributes at link time, and so the tool has to sometimes guess if it needs to create a new function that didn't already exist in the input. I don't see an obvious way around that problem. I would contend that users should be invoking their linker with the correct settings to avoid issues like this in general, so in some sense it exposes a potential problem in such builds anyway.
https://github.com/llvm/llvm-project/pull/111334#discussion_r1900369088

smithp35 reviewed (commented):
Personally I don't have any fundamental objections to a diagnostic given that the existing clang diagnostic has been kept in place. As maintainers I think the final say needs to be with @ostannard and @davemgreen . 

I'm wondering if there's any surrounding context that can used to make the warning more useful. Right now it seems to be appearing once per Subtarget. I've left a more specific comment where more context is available.

ostannard reviewed (commented):
I still think the best place to report errors like this is in the frontend or driver, so that the errors can point to exact source files, line numbers, etc. We could probably add some functions to `TargetParser` to help other frontends with these sort of compatibility checks.

However, I'm fine with this as a backup to avoid generating incorrect code when the frontend either misses a check, or is fine with the lower-quality error messages which the backend can create.

RalfJung commented:
Would it make sense to adjust the riscv backend as well for consistency? Seems a bit odd to have this check in different places for different targets.

ostannard commented:
Yes, I think moving the riscv check too would make sense.

chrisnc committed:
[ARM] Emit an error when the hard-float PCS is enabled but can't be used

Prior to this, compiling for an eabihf target with a CPU lacking
floating-point registers would silently use the soft-float ABI instead,
even though the Arm attributes section would still have
"Tag_ABI_VFP_args: VFP registers", which leads to silent ABI mismatches
at link time.

Update various tests that were using inconsistent ABI/PCS and features.

Change ARMTargetLowering::getEffectiveCallingConv from private to public
and modify it to pass through unrecognized calling conventions. Now that
ARMBaseTargetMachine::createMachineFunctionInfo calls it, leaving a
fatal error would change the behavior of IR passes that do not need to
lower calling conventions. Unrecognized calling conventions are still
treated as errors in lowering.

Fixes #110383.

smithp35 commented on llvm/test/CodeGen/ARM/arm-eabi.ll:
```diff
@@ -1,31 +1,31 @@
-; RUN: llc < %s -mtriple=arm-none-gnueabihf -meabi=5 -disable-post-ra -o - | FileCheck %s --check-prefix=CHECK-EABI
-; RUN: llc < %s -mtriple=arm-none-musleabi -meabi=5 -disable-post-ra -o - | FileCheck %s --check-prefix=CHECK-EABI
-; RUN: llc < %s -mtriple=arm-none-musleabihf -meabi=5 -disable-post-ra -o - | FileCheck %s --check-prefix=CHECK-EABI
+; RUN: llc < %s -mtriple=armv7-none-eabi -disable-post-ra -o - | FileCheck %s --check-prefix=CHECK-EABI
```
I think it would be better to make as few changes as possible and add something like `-mattr=+vfp2` to the cases that were using a `hf` suffix on the environment.

smithp35 commented on llvm/test/CodeGen/ARM/byval_struct_copy_tailcall.ll:
```diff
@@ -1,4 +1,4 @@
-; RUN: llc -mtriple thumbv7em-apple-darwin -o - < %s | FileCheck %s
+; RUN: llc -mtriple thumbv7m-apple-darwin -o - < %s | FileCheck %s
```
This is just a comment for posterity, no action needed as I wanted to check the difference between `thumbv7em-apple-darwin` and `thumbv7m-apple-darwin` and why clang didn't recognise that triple (although it does canonicalise to it internally).

Seems like for Darwin there's a default CPU which for thumbv7em is Cortex-M4 which defaults to the floating point unit being enabled, and also has a special case that hard-codes hard-float. On a non Apple triple soft-float would have been used.

smithp35 commented on llvm/test/CodeGen/ARM/subtarget-align.ll:
```diff
@@ -1,4 +1,4 @@
-; RUN: llc -mtriple=arm-linux-gnueabihf -filetype=obj <%s | llvm-objdump --triple=armv7 --no-show-raw-insn -d - | FileCheck %s
+; RUN: llc -mtriple=arm-linux-gnueabi -filetype=obj <%s | llvm-objdump --triple=armv7 --no-show-raw-insn -d - | FileCheck %s
```
This case makes me nervous. 

I was able to reproduce the failure with the following C code:
```
float f(float x, float y) {
  return x/y;
}

__attribute__((pcs("aapcs"))) __attribute__((target("arch=arm7tdmi,no-fpregs"))) float g(float x, float y) {
  return x*y;
}
```
With this patch applied:
```
clang --target=arm-linux-gnueabihf float.c -c -O2
error: Function 'g' uses the hard-float ABI, but the target lacks floating-point registers.
1 error generated.
```
With clang-22 -S output
```
f:                                      @ @f
        .fnstart
@ %bb.0:                                @ %entry
        vdiv.f32        s0, s0, s1
        bx      lr
.Lfunc_end0:
        .size   f, .Lfunc_end0-f
        .cantunwind
        .fnend
                                        @ -- End function
        .globl  g                               @ -- Begin function g
        .p2align        2
        .type   g,%function
        .code   32
g:                                      @ @g
        .fnstart
@ %bb.0:                                @ %entry
        .save   {r11, lr}
        push    {r11, lr}
        bl      __aeabi_fmul
        pop     {r11, lr}
        bx      lr
.Lfunc_end1:
```

Which looks correct to me. I think this may be that the `arm_aapcscc` is not handled by the triple in the Subtarget so it isn't accounted for in the check.

chrisnc replied:
Ack, thanks for this pointer. Indeed, the patch as-is does not account for calling conventions that differ from the default for the ABI. I'm re-working it now to deal with this. My basic idea is to check the active calling convention and determine whether it can be implemented, and this will subsume the ABI check (which determines the calling convention absent a per-function override). To do that, I needed to change `getEffectiveCallingConv` to no longer explicitly degrade `ARM_AAPCS_VFP` to `ARM_AAPCS` for lack of fpregs or thumbv1-only (otherwise checking the calling convention will conclude that it is already valid). Let me know if you think this is a valid direction to proceed. The other thing that should change is to check function calls that cannot be implemented, given the features in the caller and the calling convention of the callee.

In the process, I encountered several tests that need updating independently of any new error: https://github.com/llvm/llvm-project/pull/213588, but one was a prerequisite to enforcing that features match the calling convention (`fp16-fusedMAC.ll` which used a thumb1-only target inadvertently, but tried to use fp features).

chrisnc replied:
Updated the PR with a different approach that now handles calling convention overrides.

smithp35 reviewed (commented):
I've found one edge case that I would expect to work, but triggers the error message.

I don't think it will be common, the only reason I can think of to use it is some code that is doing run-time selection.

It is more common to have a single implementation that uses floating point registers internally, but has a hard-float and soft-float (softfp) wrappers.

chrisnc commented on llvm/test/CodeGen/ARM/constantfp.ll:
```diff
@@ -1,43 +1,43 @@
+; RUN: llc -mtriple=thumbv8m.main-none-eabihf -mattr=+execute-only -mattr=fp-armv8 -relocation-model=ropi %s -o - \
 ; RUN: | FileCheck --check-prefix=CHECK-XO-ROPI %s
 
-define arm_aapcs_vfpcc float @test_vmov_f32() {
```
The explicit calling convention here needed to be replaced by an explicit hard-float ABI in the cases where it's supported, because this test also has a `RUN: ... -mcpu=cortex-m3` which does not have an FPU, so it cannot realize the vfp calling convention.

chrisnc commented on llvm/test/DebugInfo/COFF/jump-table.ll:
```diff
@@ -231,9 +231,9 @@ declare dso_local void @f4() #2
 
 declare dso_local void @f5() #2
 
-attributes #0 = { mustprogress noinline optnone uwtable "frame-pointer"="none" "min-legal-vector-width"="0" "no-trapping-math"="true" "stack-protector-buffer-size"="8" "target-cpu"="x86-64" "target-features"="+cx8,+fxsr,+mmx,+sse,+sse2,+x87" "tune-cpu"="generic" }
```
This test has `RUN` lines for Arm, so these attributes were wrong and accidentally deleting the Arm FPU features.

chrisnc commented on clang/test/CodeGenCXX/arm-pcs.cpp:
```diff
@@ -6,6 +6,13 @@
 // RUN: %clang -mfloat-abi=softfp --target=armv7-unknown-linux-gnueabi -O3 -S -o - %s | FileCheck %s -check-prefixes=SOFTFP,CHECK
 // RUN: %clang -mfloat-abi=soft --target=armv7-unknown-linux-gnueabi -O3 -S -o - %s | FileCheck %s -check-prefixes=SOFT,CHECK
```
`-mfloat-abi=soft` also does `-fpregs`, so the test code as written can't implement `__attribute__((pcs("aapcs-vfp")))` for this `RUN` line. This was the most straightforward way I could see to address, without sacrificing other things this test wants to do, or tearing it up. Open to other ideas.

chrisnc commented on llvm/test/CodeGen/ARM/byval_struct_copy_tailcall.ll:
```diff
@@ -1,65 +1,69 @@
-; RUN: llc -mtriple thumbv7em-apple-darwin -o - < %s | FileCheck %s
```
I regenerated the checks for this one because `-darwin` implies a hard-float ABI when it's not needed here, but that also changed some of the emitted assembly in small ways. This test does not depend on darwin or hard-float.

smithp35 reviewed (commented):
I think I'm down to one suggestion for not using a friend as it looks like we can use a free function instead. That is more personal preference though so if it turns out to be a pain we can keep the friend.

Otherwise this is looking good to me. Thanks for all the hard work in updating the tests.

smithp35 reviewed (approved):
Thanks for the updates. This looks good to me. I remember that you don't have merge permissions. If there are no objections I'll merge this on Monday next week. That way if there are some unforeseen problems with projects not caught by precommit, we'll have more time to react.

statham-arm commented:
Hi @chrisnc,

This new diagnostic caused some build failures in our Arm embedded toolchain. We build some embedded C/C++ libraries intended to run on a platform with hardware FP, but because they're embedded, the code begins running when the FPU is not enabled in hardware yet. So we tagged one of our initial setup functions with the `target("no-fpregs")` attribute to make sure it didn't accidentally use the FP registers:
```c++
[[gnu::target("no-fpregs")]] void setup() { ... }
```
But now this causes a compile error, because the compiler is running with `-mfloat-abi=hard`, and `setup()` isn't allowed to have the hard-float ABI and also not use the FP registers.

However, its function type is `void(void)`, which works the same with or without the hard-float ABI! The diagnostic is complaining that it doesn't have the same _formal_ ABI, even though it makes no difference in this case. We have to work around it by tagging the function with the `pcs("aapcs")` attribute too.

Another case of this occurs when the setup function calls a subroutine to write a system register. That subroutine must also be formally marked as `pcs("aapcs")` to avoid this new diagnostic. Again that function has no floating-point values in its signature. But not only that: it's also marked `[[clang::always_inline]]`, which ought to mean that its PCS _does not matter_, because no function call to it is ever emitted in output code!

What do you think? Would it make sense to suppress this diagnostic in one or both of the cases where
* the function's type is such that the two ABIs would agree anyway
* the function is marked always-inline, so that its ABI is irrelevant?

Separately, here's an oddity in the error reporting. Consider this input code:
```c++
__attribute__((target("no-fpregs"), pcs("aapcs"))) int f(void) {
    extern int g(void);
    return g() + 1;
}
```
Here, `f` is fine: it's not using the FP registers, but that's OK, it's also marked with the soft-float ABI. But `g` isn't, so the call to `g` provokes your new diagnostic. However, the diagnostic is localised to the start of `f`:
```
$ clang --target=armv8a-none-eabi -mfloat-abi=hard -c zz.c
zz.c:1:56: error: call to 'g' expects a hard-float calling convention, but
      floating-point registers are unavailable
    1 | __attribute__((target("no-fpregs"), pcs("aapcs"))) int f(void) {
      |                                                        ^
1 error generated.
```
If `f` is a long function, it might not be easy to track down _where_ in it I've accidentally called `g`. Shouldn't the error message point at the objectionable function call, not the start of the containing function?

chrisnc commented:
@statham-arm thanks for the report.

For the call to `setup`, how is it that the caller has `fpregs` and `aapcs_vfp`, if the callee does not, if the premise is that fp regs are not available at this point? I think it's reasonable to adjust both the feature and the calling convention in this instance, as one depends on the other.

As for the diagnostic pointing at the caller rather than the callee, I will look at improving that.

I'll also look into the always-inline case and whether that could be improved. My first thought there is that even with `always_inline`, semantically the calling convention and features are still relevant if they differ between caller and callee, and it's still important to flag mismatches (in this case, the system register function would have the vfp calling convention, but its caller is missing the fpregs feature).

Whether and how to suppress the diagnostic when the mismatch "doesn't matter" is tricky, depending on when different parts of the backend use these attributes. In a previous version of this PR, that was attempted but it failed because of how subtargets are created, but perhaps the current implementation is more amenable to something like this. Perhaps also there is a world where functions whose nominal calling convention is aapcs_vfp, but whose arguments are such that aapcs is equivalent, are actually given an aapcs calling convention, but this may be a pretty extensive change.

chrisnc commented:
@statham-arm I tried the example of the weird diagnostic. If you pass `-g`, the call site is shown in the output as expected. This is because without `-g` there are no `!dbg` attachments on the IR at the call site to use in the diagnostic, only the function itself. I agree the message is confusing in the case where this is not present though.

```bash
$ build/bin/clang --target=armv8a-none-eabi -mfloat-abi=hard -c zz.c -g
zz.c:3:12: error: call to 'g' expects a hard-float calling convention, but floating-point
      registers are unavailable
    3 |     return g() + 1;
      |            ^
1 error generated.
```

chrisnc commented:
I created https://github.com/llvm/llvm-project/pull/217693 to improve the diagnostic for the call case.
